### PR TITLE
KAFKA-6321: Consolidate calls to KafkaConsumer's `beginningOffsets()` and `endOffsets()` in ConsumerGroupCommand

### PR DESCRIPTION
**Pull request by vahidhashemian**
### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala [hunk 1]
```diff
@@ -583,16 +584,23 @@ object ConsumerGroupCommand extends Logging {
...
     }
 
-    protected def getLogEndOffset(topicPartition: TopicPartition): LogOffsetResult = {
-      val offsets = getConsumer.endOffsets(List(topicPartition).asJava)
-      val logStartOffset = offsets.get(topicPartition)
-      LogOffsetResult.LogOffset(logStartOffset)
+    protected def getLogEndOffsets(topicPartitions: Seq[TopicPartition]): Map[TopicPartition, LogOffsetResult] = {
+      val offsets = getConsumer.endOffsets(topicPartitions.asJava)
+      topicPartitions.map { topicPartition =>
+        val logStartOffset = offsets.get(topicPartition)
```
Comment (tedyu) [inline core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: Should this variable be renamed to match the method name ?

The method is getLogEndOffsets.

Comment (vahidhashemian) [reply core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: Yes, thanks for catching it. I'll submit an update.

**Comment by vahidhashemian** [comment]
@becketqin This PR addresses the JIRA you opened. I'd appreciate your review when you get a chance. Thanks.

**Comment by becketqin** [comment]
Thanks for the patch. I am a little busy today. Will review tomorrow.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala [hunk 2]
```diff
@@ -274,7 +274,10 @@ object ConsumerGroupCommand extends Logging {
 
     protected def opts: ConsumerGroupCommandOptions
 
-    protected def getLogEndOffset(topicPartition: TopicPartition): LogOffsetResult
+    protected def getLogEndOffset(topicPartition: TopicPartition): LogOffsetResult =
+      getLogEndOffsets(Seq(topicPartition)).get(topicPartition).getOrElse(LogOffsetResult.Ignore)
```
Comment (becketqin) [inline core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:291]: Could just be ```getLogEndOffsets(Seq(topicPartition)).getOrElse(topicPartition, LogOffsetResult.Ignore)```

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala [hunk 3]
```diff
@@ -274,7 +274,10 @@ object ConsumerGroupCommand extends Logging {
 
     protected def opts: ConsumerGroupCommandOptions
 
-    protected def getLogEndOffset(topicPartition: TopicPartition): LogOffsetResult
+    protected def getLogEndOffset(topicPartition: TopicPartition): LogOffsetResult =
```
Comment (becketqin) [inline core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:290]: It seems this method is still used in a loop in line 706 and line 771. Should we also replace them with batch call?

**Review by becketqin — COMMENTED** [review]
Thanks for the patch. Please see some comments below.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala (hunk 3 above)
Comment (vahidhashemian) [reply core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:290]: You're right. I'll try to consolidate this in the next commit. Thanks for catching it.

**Review by vahidhashemian — COMMENTED** [review]
@becketqin, thanks for reviewing the PR. I tried to address your comments in the new commit.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala [hunk 4]
```diff
@@ -583,16 +585,23 @@ object ConsumerGroupCommand extends Logging {
...
+        val logEndOffset = offsets.get(topicPartition)
+        topicPartition -> LogOffsetResult.LogOffset(logEndOffset)
+      }.toMap
     }
 
-    protected def getLogStartOffset(topicPartition: TopicPartition): LogOffsetResult = {
-      val offsets = getConsumer.beginningOffsets(List(topicPartition).asJava)
-      val logStartOffset = offsets.get(topicPartition)
-      LogOffsetResult.LogOffset(logStartOffset)
+    protected def getLogStartOffset(topicPartition: TopicPartition): LogOffsetResult =
```
Comment (becketqin) [inline core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: It seems this method is no longer used. Can we just delete it?

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala [hunk 5]
```diff
@@ -583,16 +585,23 @@ object ConsumerGroupCommand extends Logging {
...
+
+    protected def getLogStartOffsets(topicPartitions: Seq[TopicPartition]): Map[TopicPartition, LogOffsetResult] = {
+      val offsets = getConsumer.beginningOffsets(topicPartitions.asJava)
+      topicPartitions.map { topicPartition =>
+        val logStartOffset = offsets.get(topicPartition)
+        topicPartition -> LogOffsetResult.LogOffset(logStartOffset)
+      }.toMap
     }
 
     protected def getLogTimestampOffset(topicPartition: TopicPartition, timestamp: java.lang.Long): LogOffsetResult = {
```
Comment (becketqin) [inline core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: This method is called when resetting offsets by timestamp and duration. It seems we are still doing the reset partition by partition. Can we also batch them?

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala [hunk 6]
```diff
@@ -769,19 +781,27 @@ object ConsumerGroupCommand extends Logging {
       }
     }
 
-    private def checkOffsetRange(topicPartition: TopicPartition, offset: Long) = {
-      getLogEndOffset(topicPartition) match {
-        case LogOffsetResult.LogOffset(endOffset) if offset > endOffset =>
-          warn(s"New offset ($offset) is higher than latest offset. Value will be set to $endOffset")
-          endOffset
+    private def checkOffsetRange(requestedOffsets: Map[TopicPartition, Long]) = {
```
Comment (becketqin) [inline core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: Nit: perhaps we should change the method name to checkOffsetsRange(), with plural form.

**Review by becketqin — COMMENTED** [review]
@vahidhashemian Thanks for updating the patch. Happy new year! I left a few more comments.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala (hunk 4 above)
Comment (vahidhashemian) [reply core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: That's true. I'll remove it in the next commit. Thanks!

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala (hunk 5 above)
Comment (vahidhashemian) [reply core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: Yes, I'll do that in the next commit.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala (hunk 6 above)
Comment (vahidhashemian) [reply core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: Sure, will update this too.

**Review by vahidhashemian — COMMENTED** [review]
@becketqin Happy new year! Thanks for another round of feedback. I tried to address them in the new commit.

**Comment by vahidhashemian** [comment]
@becketqin any other comment/feedback on this PR? Thanks.

**Comment by becketqin** [comment]
@vahidhashemian Sorry I missed the PR update. Will look at it later today.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala [hunk 7]
```diff
@@ -583,26 +585,34 @@ object ConsumerGroupCommand extends Logging {
...
+    protected def getLogTimestampOffsets(topicPartitions: Seq[TopicPartition], timestamp: java.lang.Long): Map[TopicPartition, LogOffsetResult] = {
       val consumer = getConsumer
-      consumer.assign(List(topicPartition).asJava)
-      val offsetsForTimes = consumer.offsetsForTimes(Map(topicPartition -> timestamp).asJava)
-      if (offsetsForTimes != null && !offsetsForTimes.isEmpty && offsetsForTimes.get(topicPartition) != null)
-        LogOffsetResult.LogOffset(offsetsForTimes.get(topicPartition).offset)
+      consumer.assign(topicPartitions.asJava)
+      val offsetsForTimes = consumer.offsetsForTimes(topicPartitions.map(_ -> timestamp).toMap.asJava)
+
+      if (offsetsForTimes == null || offsetsForTimes.isEmpty)
```
Comment (becketqin) [inline core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: KafkaConsumer.offsetsForTimes() will always return a non-empty map as long as the initial partition list is not empty. If the offset of a particular partition is not available, the returned map will have that partition mapped to null.

Perhaps we can do the following here:
1. use offsetsForTimes() to get the result for all the specified partitions.
2. If there are partitions with null value in the map returned in step 1, add all those unsuccessful partitions to another set.
3. use getEndOffsets() to get the log end offsets for the unsuccessful partitions in step 2 and map those partitions to their log end offsets.
4. return the combined result from step 1 and 3.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala [hunk 8]
```diff
@@ -690,68 +700,73 @@ object ConsumerGroupCommand extends Logging {
...
-          (topicPartition, new OffsetAndMetadata(newOffset))
+        val requestedOffsets = resetPlan.keySet.map { topicPartition =>
+          (topicPartition, resetPlan(topicPartition).offset())
         }.toMap
+        checkOffsetsRange(requestedOffsets).map {
+          case (topicPartition, newOffset) => (topicPartition, new OffsetAndMetadata(newOffset))
+        }
       } else if (opts.options.has(opts.resetToCurrentOpt)) {
         val currentCommittedOffsets = adminClient.listGroupOffsets(groupId)
         partitionsToReset.map { topicPartition =>
```
Comment (becketqin) [inline core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: Can we also batch the requests here? Perhaps we can add all the partitions without a committed offsets to a separate set and invoke the getLogEndOffsets() for those partitions.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala [hunk 9]
```diff
@@ -769,19 +784,27 @@ object ConsumerGroupCommand extends Logging {
...
+            case Some(LogOffsetResult.LogOffset(startOffset)) if offset < startOffset =>
+              warn(s"New offset ($offset) is lower than earliest offset for topic partition $topicPartition. Value will be set to $startOffset")
+              startOffset
 
-          case _ => offset
-        }
+            case _ => offset
+          }
+
+          case None => // the control should not reach here
```
Comment (becketqin) [inline core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:822]: If the control should not reach here, maybe we can throw an IllegalStateException?

**Review by becketqin — COMMENTED** [review]
@vahidhashemian Thanks for updating the patch. Made another pass and left some comments.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala (hunk 7 above)
Comment (vahidhashemian) [reply core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: Thanks, I made this change in the latest commit. It should save some cpu cycles.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala (hunk 8 above)
Comment (vahidhashemian) [reply core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala]: Sure, that makes sense too. I've done that in the latest commit.

**Review by vahidhashemian — COMMENTED** [review]
@becketqin Thanks for another look at this PR, and the suggestions. I tried to address them (additional comments inline) in the recent commit.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala [hunk 10]
```diff
@@ -769,16 +768,22 @@ object ConsumerGroupCommand extends Logging {
...
-              case _ => CommandLineUtils.printUsageAndDie(opts.parser, s"Error getting ending offset of topic partition: $topicPartition")
-            }
-          )
+        val (partitionsToResetWithCommittedOffset, partitionsToResetWithoutCommittedOffset) =
+          partitionsToReset.partition(currentCommittedOffsets.keySet.contains(_))
+
+        val preparedOffsetsForParititionsWithCommittedOffset = partitionsToResetWithCommittedOffset.map { topicPartition =>
+          (topicPartition, new OffsetAndMetadata(currentCommittedOffsets.get(topicPartition) match {
+            case Some(offset) => offset
+            case _ => throw new IllegalStateException(s"Expected a valid current offset for topic partition: $topicPartition")
```
Comment (becketqin) [inline core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:790]: Currently we print the help message and exit. Is this change intentional?

**Review by becketqin — COMMENTED** [review]
Thanks for the latest patch. LGTM. Just one minor comment.

File: core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala (hunk 10 above)
Comment (vahidhashemian) [reply core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:790]: This is another part of the code that should be reachable (because at this point we're certain `currentCommittedOffsets.get(topicPartition)` is not `None`). I used this instead of doing a `currentCommittedOffsets.get(topicPartition).get` above. If you prefer that one instead, or can think of a better way to go about it (that avoids using `.get` on an `Option`) please let me know and I'll make the change. Thanks.

Comment (becketqin) [reply core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:790]: Ah, this is for the partitions with committed offsets. This is fine.

**Comment by becketqin** [comment]
@vahidhashemian It looks there is a minor conflict when I was trying to check in the patch. Do you have time to rebase the patch? Thanks.

**Comment by vahidhashemian** [comment]
@becketqin Thanks for letting me know. The rebase is complete.

**Comment by becketqin** [comment]
@vahidhashemian Thanks for the patch. I just merged it to the trunk. 
BTW, I found another related issue when testing the patch and created KAFKA-6489.

**Comment by vahidhashemian** [comment]
@becketqin, thanks! Let me know if you want me to work on that too.

**Comment by becketqin** [comment]
@vahidhashemian You are welcome. I have submitted a patch for KAFKA-6489. It is a small fix.
